### ISSUE TITLE
Java: Audit the Otel implementation of Baggage using Baggage Test Suite

### ISSUE THREAD

**Issue by kalyanaj** [w3c/baggage#138]
This is based on the exit criteria for the Candidate Recommendation State: "During the Candidate Recommendation stage, the Working Group will audit:
the [OpenTelemetry implementations of Baggage](https://opentelemetry.io/docs/specs/otel/baggage/api/) to make sure they are following the W3C Baggage spec."

Target Milestone: Oct 31 2024 so that we can proceed with the next steps for getting Baggage into recommendation state.

**Comment by basti1302** [comment]
I took our own test suite (https://github.com/w3c/baggage/blob/main/test/test_baggage.py) and converted it to Java here: https://github.com/basti1302/opentelemetry-java/blob/w3c-baggage-issue-138-audit-otel-java-sdk/api/all/src/test/java/io/opentelemetry/api/baggage/propagation/W3cSpecTest.java

**TL;DR: I think the OTel Java SDK implementation of baggage is not compliant with this spec when properties/metadata are involved.**

This is due to one fundamental difference between the OTel baggage spec and this spec. The W3C spec allows additional/optional properties on list members. There properties itself can also be key-value pairs (like list members). Like so:

```
baggage: key1=value1;property1, key2=value2;property2=propertyValue2
```

which translates to two list members:
* `key1: value1` with a flag-like property `property1` (without value)
* `key2: value2` with a property `property2: propertyValue2`.

In contrast the OTel baggage spec supports additional properties on list members under the term "metadata", but metadata is *one opaque string*, and it is not interpreted or parsed. 
 
* Test cases that only involve simple key-value pairs (without properties) all work as expected.
* Test cases where the properties-part of a key-value pair does not include white space and does not include key-value properties also work as expected.

But when the properties-part of a list member includes OWS or is a key value pair (and thus includes the equals character), the OTel Java SDK will
* treat the whole property as one opaque string
* percent-encode it when propagating it downstream.

Consider the following baggage header:

```
baggage: SomeKey=SomeValue;ValueProp \t = \t PropVal
```

We would expect this to be understood as one list member, with key `SomeKey`, value `SomeValue` and one property `ValueProp: PropVal`.

However, the Java OTel SDK will propagate this downstream as follows:

```
baggage: SomeKey=SomeValue;ValueProp%20%09%20%3D%20%09%20PropVal"
```

That is, the inner whitespace and the equals sign in the property are being percent-encoded.

Now, when this header would be parsed by a spec-compliant baggage implementation, this would probably be parsed as follows: One list member, with key `SomeKey`, value `SomeValue` and one property which no longer has key-value shape but is now a flag-like property `ValueProp%20%09%20%3D%20%09%20PropVal` (or the percent-decoded variant of this, but still as a single string).

At least this is what our reference implementation would do. To verify this, I added this test case locally on my end:
```
    def test_parse_property_percent_encoded(self):
        baggage = Baggage().from_string(
            "SomeKey=SomeValue;ValueProp%20%09%20%3D%20%09%20PropVal")
        self.assertEqual(len(baggage.entries), 1)
        entry1 = baggage.entries[0]
        self.assertEqual(entry1.key, "SomeKey")
        self.assertEqual(entry1.value, "SomeValue")
        self.assertEqual(entry1.properties[0].key, "ValueProp%20%09%20%3D%20%09%20PropVal")
        self.assertEqual(entry1.properties[0].value, None)
```

Additional note: AFAIC, our spec does not spell out whether or not properties are subject to percent encoding. This is only specified for values. We should spell this out explicitly (and I assume at least property values should also be subject to percent encoding?).

**Comment by dyladan** [comment]
Hmm. I remember when this decision was made it was specifically because the metadata had no meaning or purpose yet. When OTel was implementing the draft spec, we didn't know what metadata might turn into so we decided to treat it as opaque to leave options open for the future. I will have to test this specifically, but I think the JS implementation doesn't do the percent encoding trick, but just passes it _directly_ unchanged.

**Comment by basti1302** [comment]
We discussed this in the working group meeting yesterday and came to the following consensus: We will file this as an implementation bug with the Java OTel SDK. Neither our spec nor the OTel baggage spec mandates to percent-encode properties/metadata. In fact, the OTel spec mandates to treat metadata as an opaque string (and not process it at all). We will also check other OTel SDKs whether they exhibit similar behavior.

Tangentially related, we are also considering changing the MUST for percent encoding values of list-members. Instead of prescribing a specific processing, we probably should only specify the allowed characters without mandating any specific mechanism for encoding not-allowed characters. (Percent-encoding could be mentioned as a non-normative hint as one possible option.)

**Comment by basti1302** [comment]
The Java OTel SDK issue has been filed here: https://github.com/open-telemetry/opentelemetry-java/issues/6771

**Comment by yurishkuro** [comment]
Hold on - the spec absolutely must mandate encoding, otherwise the consumer never knows how to interpret the data, whether to decode it or not. We had this ambiguity in jaeger baggage format and it was very painful to handle

**Comment by basti1302** [comment]
@yurishkuro I created a new issue for this part of the discussion -> https://github.com/w3c/baggage/issues/145.

**Comment by SergeyKanzhelev** [comment]
Java implementation is compliant except this issue: https://github.com/open-telemetry/opentelemetry-java/issues/6771

Closing this audit task
